loopchain/blockchain/vote.py

```diff
@@ -69,39 +66,44 @@ def votes(self):
     def target_hash(self):
         return self.__target_hash
 
-    @staticmethod
-    def __make_vote_init(audience):
+    def __make_vote_init(self, audience):
         if not audience:
             return None
 
         vote_init = {}
+        self.__result_count_list = {}
         if isinstance(audience, PeerManager):
-            for group_id in list(audience.peer_list.keys()):
-                if group_id == conf.ALL_GROUP_ID:
-                    continue
-                vote_init[group_id] = {}
-                for peer_id in list(audience.peer_list[group_id].keys()):
-                    vote_init[group_id][peer_id] = []
-        else:
-            for peer_id in audience:
-                vote_init[audience[peer_id].group_id] = {}
-                vote_init[audience[peer_id].group_id][peer_id] = []
+            audience = list(audience.peer_list[conf.ALL_GROUP_ID].keys())
+
+        for peer_id in audience:
+            vote_init[peer_id] = {}
 
         logging.debug("vote_init: " + str(vote_init))
         return vote_init
 
-    @staticmethod
-    def __parse_vote_sign(vote_sign):
-        """서명된 vote로 부터 투표 결과를 추출한다."""
+    def __parse_vote_sign(self, vote_sign):
+        """Derive result of vote from vote_sign."""
 
         return vote_sign
 
-    def add_vote(self, group_id, peer_id, vote_sign):
-        if group_id not in self.__votes.keys():
+    def add_vote(self, peer_id, vote_sign):
+        if peer_id not in self.__votes.keys():
             return False
-        if peer_id not in self.__votes[group_id].keys():
+
+        result = self.__parse_vote_sign(vote_sign)
+
+        if self.__votes[peer_id]:
+            logging.debug(f"This peer already votes.\n"
+                          f"old:({peer_id} to {self.__votes[peer_id][0]})\nnew:({peer_id} to {result}) ")
             return False
-        self.__votes[group_id][peer_id] = (self.__parse_vote_sign(vote_sign), vote_sign)
+        else:
+            self.__votes[peer_id] = (result, vote_sign)
+
+        if result in self.__result_count_list:
```

Comment:
You can remove this if statement using `defaultdict`.
Or How about using `Counter`?

Reply:
Thank you for your good suggestion ! I'll try it :)